scripts/build_ses.sh

```diff
@@ -0,0 +1,284 @@
+    exit 1
+fi
+
+if [[ "$1" =~ $ip_cidr_regex ]]
+then
+    public_network=$1
+    shift
+    if [[ "$1" =~ $ip_cidr_regex ]]
+    then
+        cluster_network=$1
+        shift
+    fi
+fi
+master=$1
+shift
+minions=($@)
+
+function run_ssh_cmd() {
+    host=$1
+    shift
+    ssh $host -C "$@"
+    if [ $? -gt 0 ]
+    then
+        echo "Ssh command '$@' failed on host '$host'"
+        exit 2
+    fi
+}
+
+function salt_setup() {
+    # install salt master
+    run_ssh_cmd $master "zypper -n in deepsea"
+    run_ssh_cmd $master "systemctl enable salt-master.service"
+    run_ssh_cmd $master "systemctl start salt-master.service"
+    run_ssh_cmd $master "echo 'master: $master' > /etc/salt/minion.d/master.conf"
+    run_ssh_cmd $master "sudo systemctl enable salt-minion.service"
+    run_ssh_cmd $master "sudo systemctl start salt-minion.service"
+    if ssh $master "[ ! -f /srv/pillar/ceph/deepsea_minions.sls.bak ]"
+    then
+        run_ssh_cmd $master "sudo cp /srv/pillar/ceph/deepsea_minions.sls /srv/pillar/ceph/deepsea_minions.sls.bak"
+        cat <<EOF | ssh $master "cat > /srv/pillar/ceph/deepsea_minions.sls"
+# Using the deepsea grain 'G@deepsea:*' wasn't working, so setting to all nodes.
+# original file saved to /srv/pillar/ceph/deepsea_minions.sls.bak
+deepsea_minions: '*'
+EOF
+    fi
+
+    # install salt minion on the minions
+    for minion in "${minions[@]}"
+    do
+        run_ssh_cmd $minion "sudo zypper -n in salt-minion"
+        run_ssh_cmd $minion "sudo systemctl enable salt-minion.service"
+        run_ssh_cmd $minion "echo 'master: $master' > /etc/salt/minion.d/master.conf"
+        run_ssh_cmd $minion "sudo systemctl start salt-minion.service"
+    done
+
+    # Back on master, accept all the minions salt keys
+    run_ssh_cmd $master "salt-key --accept-all -y"
+    echo -n "waiting on keys (wait 20 times in 2 sec increments)."
+    for x in $(seq 20)
+    do
+        if [ $(run_ssh_cmd $master "salt-key -l acc |grep -cv 'Accepted Keys:'") -ne $(( ${#minions[@]} + 1 )) ]
+        then
+            echo '.'
+            run_ssh_cmd $master "salt-key --accept-all -y 2> /dev/null"
+            sleep 2
+        else
+            echo 'done'
+            break
+        fi
+    done
+    if [ $(run_ssh_cmd $master "salt-key -l acc |grep -cv 'Accepted Keys:'") -ne $(( ${#minions[@]} + 1 )) ]
+    then
+        echo 'failed'
+        exit 1
+    fi
+}
+
+function rnd_sel_nodes() {
+    num=$1
+    shift
+    local nodes
+    nodes=($@)
+
+    if [ $num -ge ${#nodes[@]} ]
+    then
+        echo ${nodes[@]}
+        return
+    fi
+
+    result=()
+    while [ $num -gt 0 ]
+    do
+        rnd=$(( RANDOM % ${#nodes[@]} ))
+        result=(${result[@]} ${nodes[$rnd]})
+        unset nodes[$rnd]
+        nodes=( ${nodes[@]} )
+        let "num-=1"
+    done
+
+    echo ${result[@]}
+}
+
+function generate_policy_config() {
+    pconf="/srv/pillar/ceph/proposals/policy.cfg"
```

Comment:
Right now policy.cfg is generated incorrectly.

You are missing the role-mon/stack and role-mon/cluster for the node that is both master and a minion.

So for example node1 is a minion and a master of itself; node2 is a minion' node3 is a minion.
Then currently only entries for node2 and node3 will be generated in policy.cfg.

Reply:
oh really, I'll take a look, thanks!

Reply:
Done.
Originally I read somewhere to not make the salt master a monitor, which is why it was left off.. but this is 1 for testing and 2, I now can't find that statement. So now have just added the master as a potential candidate to any role, so smaller lonely nodes (say 3) would still work.

Reply:
For the record: http://beta.suse.com/private/SUSE-Storage-Beta/doc/ses-manual_en/ceph.install.saltstack.html says: "Do not run Ceph related services on the Salt master."